rust/src/storage/file.rs

```diff
@@ -58,26 +65,110 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
-
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
-        }
+        let tmp_file = self
+            .root
+            .join("_tmp_writes")
+            .join(Uuid::new_v4().to_string());
 
-        // use `OpenOptions` with `create_new` to create the file handle.
-        // this will force ErrorKind::AlreadyExists if the file already exists.
-        let mut f = fs::OpenOptions::new()
-            .create_new(true)
-            .write(true)
-            .open(path)
-            .await
-            .map_err(|e| match e.kind() {
-                std::io::ErrorKind::AlreadyExists => StorageError::AlreadyExists(path.to_string()),
-                _ => StorageError::Io { source: e },
-            })?;
+        create_parent_dirs(&tmp_file).await?;
+        create_parent_dirs(Path::new(path)).await?;
 
+        let mut f = fs::File::create(&tmp_file).await?;
         f.write(obj_bytes).await?;
 
+        rename(tmp_file.to_str().unwrap(), path)?;
+
         Ok(())
     }
 }
+
+async fn create_parent_dirs(p: &Path) -> std::io::Result<()> {
+    if let Some(d) = p.parent() {
+        fs::create_dir_all(d).await?;
+    }
+    Ok(())
+}
+
+#[cfg(target_os = "linux")]
+const RENAME_NOREPLACE: c_uint = 1;
+
+#[cfg(target_os = "macos")]
+const RENAME_EXCL: c_uint = 4;
+
+extern "C" {
+    #[cfg(target_os = "linux")]
+    fn renameat2(
+        olddirfd: c_int,
+        oldpath: *const c_char,
+        newdirfd: c_int,
+        newpath: *const c_char,
+        flags: c_uint,
+    ) -> c_int;
+
+    #[cfg(target_os = "macos")]
+    fn renamex_np(from: *const c_char, to: *const c_char, flags: c_uint) -> c_int;
+}
+
+fn rename(from: &str, to: &str) -> Result<(), StorageError> {
+    let from_s = CString::new(from).unwrap();
+    let to_s = CString::new(to).unwrap();
```

Comment:
is it safe to call unwrap here?
`?` operator didn't work since the `From` for returned `Err` is missing. Should I leave it as is or create new `From` impl

Reply:
In theory it's safe, but I think it's better to propagate the error back to caller. You don't have to implement `From` trait for `NulError`, you can also just apply `map_err` with a named or anonymous function.

Reply:
Done that, howevevr I've introduced new `Generic` error with plain text https://github.com/delta-io/delta-rs/pull/126/files#diff-8e70f44f4dac2ac3e0e0e1d3d3ebfcad6032c979eab15a5b5c97984753cfd546R163, wdyt?

Reply:
looks good 👍